Make build profile project detection also set build options

YeldhamDev opened the pull request:
Currently, the build profile editor only updates the used classes when attempting to detect from the project. This PR makes so that the other settings are also updated from the project information with the following.

- Checking project settings.
- Checking classes that require them.
- And last but definitely not least, import settings. This one required me to add an extra functionality called `get_build_dependencies()`, that works like `get_classes_used()` but is geared towards resource importers. And just like the latter, the former is also exposed to scripting.

However, that are a couple of build options that I couldn't find a way to detect, because as far as I'm aware, there are no settings that could imply if the project is using them or not:
- ~X11/Wayland~
- SIL Graphite Fonts

This PR also fixes some issues necessary to make the feature to work properly:
- Fix error when scanning resource files.
- Fix loading profile files doing nothing.
- Hardcode some classes that are necessary for the engine to function.

**Sponsored By:** 🐺 Lone Wolf Technology / 🍀 W4 Games.

YeldhamDev committed:
Make build profile project detection also set build options

YeldhamDev commented:
@Riteo Yep, missed that setting. If it's set to `default`, both are now selected, otherwise, it will be the one you picked.

YeldhamDev commented:
This PR is ready for review.

Calinou commented:
I get this error when compiling a release export template with a profile generated from the 3D platformer demo:

```
main/main.cpp:188:8: error: unknown type name 'XRServer'
  188 | static XRServer *xr_server = nullptr;
      |        ^
1 error generated.
Compiling modules/gltf/extensions/gltf_document_extension.cpp ...
scons: *** [bin/obj/main/main.macos.template_release.arm64.o] Error 1
scons: building terminated because of errors.
INFO: Time elapsed: 00:00:07.16
```

This doesn't occur with a profile generated from `master`.

Profile files: [build_profiles.zip](https://github.com/user-attachments/files/19590759/build_profiles.zip)

SCons command:

```
scons -j16 progress=no target=template_release optimize=speed build_profile=path/to/profile.gdbuild
```

YeldhamDev commented:
@Calinou Try again, please. Seem that it was just a missing `#ifndef`.

akien-mga commented:
> I get this error on macOS when compiling a release export template with a profile generated from the [3D platformer demo](https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer):
> 
> ```
> main/main.cpp:188:8: error: unknown type name 'XRServer'
>   188 | static XRServer *xr_server = nullptr;
>       |        ^
> 1 error generated.
> Compiling modules/gltf/extensions/gltf_document_extension.cpp ...
> scons: *** [bin/obj/main/main.macos.template_release.arm64.o] Error 1
> scons: building terminated because of errors.
> INFO: Time elapsed: 00:00:07.16
> ```

This is being fixed in #104978.

Calinou commented:
I'm getting a different build error after rebasing on top on `master` on Linux:

```
Compiling scene/resources/scu/scu_scene_resources_4.gen.cpp ...
In file included from editor/scu/scu_editor_2.gen.cpp:2:
./editor/editor_build_profile.cpp: In member function 'void EditorBuildProfileManager::_detect_from_project()':
./editor/editor_build_profile.cpp:917:62: error: cannot convert 'List<StringName>*' to 'LocalVector<StringName>&'
  917 |                 ClassDB::get_inheriters_from_class(hc_class, &inheriters);
      |                                                              ^~~~~~~~~~~
      |                                                              |
      |                                                              List<StringName>*
In file included from ./core/object/ref_counted.h:33,
                 from ./core/object/script_instance.h:33,
                 from ./core/object/gdvirtual.gen.inc:4,
                 from ./scene/gui/control.h:34,
                 from ./scene/gui/container.h:33,
                 from ./scene/gui/box_container.h:33,
                 from ./editor/editor_autoload_settings.h:33,
                 from ./editor/editor_autoload_settings.cpp:31,
                 from editor/scu/scu_editor_2.gen.cpp:1:
./core/object/class_db.h:314:99: note:   initializing argument 2 of 'static void ClassDB::get_inheriters_from_class(const StringName&, LocalVector<StringName>&)'
  314 |         static void get_inheriters_from_class(const StringName &p_class, LocalVector<StringName> &p_classes);
      |                                                                          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
scons: *** [bin/obj/editor/scu/scu_editor_2.gen.linuxbsd.editor.x86_64.o] Error 1
```

YeldhamDev commented:
@Calinou Alright, the problem was that class extraction from binary resources was broken (like almost everything about this feature before this PR. :melting_face:). I have now fixed it, please try again.

Calinou commented:
> That's not right, it manages to find `GridMap` on my end. Are you deleting the cache before re-detecting?

It works after removing `.godot/` :slightly_smiling_face:
The project works just fine with the custom build profile now.

These messages are still printed on startup though:

```
WARNING: Property not found: 'editor/movie_writer/mix_rate'.
     at: get_setting_with_override (core/config/project_settings.cpp:405)
WARNING: Property not found: 'editor/movie_writer/speaker_mode'.
     at: get_setting_with_override (core/config/project_settings.cpp:405)
WARNING: Property not found: 'editor/movie_writer/mjpeg_quality'.
     at: get_setting_with_override (core/config/project_settings.cpp:405)
```

Calinou reviewed (approved):
Works great now!

AThousandShips commented on editor/editor_build_profile.cpp:
```diff
@@ -210,16 +474,38 @@ String EditorBuildProfile::get_build_option_description(BuildOption p_build_opti
 }
 
+LocalVector<EditorBuildProfile::BuildOption> EditorBuildProfile::get_build_option_dependencies(BuildOption p_build_option) {
+	ERR_FAIL_INDEX_V(p_build_option, BUILD_OPTION_MAX, {});
```
```suggestion
	ERR_FAIL_INDEX_V(p_build_option, BUILD_OPTION_MAX, LocalVector<EditorBuildProfile::BuildOption>());
```
Please be explicit about return types

AThousandShips commented on editor/editor_build_profile.cpp:
```diff
@@ -210,16 +474,38 @@ String EditorBuildProfile::get_build_option_description(BuildOption p_build_opti
+}
+
+HashMap<String, LocalVector<Variant>> EditorBuildProfile::get_build_option_settings(BuildOption p_build_option) {
+	ERR_FAIL_INDEX_V(p_build_option, BUILD_OPTION_MAX, {});
```
```suggestion
	ERR_FAIL_INDEX_V(p_build_option, BUILD_OPTION_MAX, HashMap<String, LocalVector<Variant>>());
```

AThousandShips commented on editor/editor_build_profile.cpp:
```diff
@@ -210,16 +474,38 @@ String EditorBuildProfile::get_build_option_description(BuildOption p_build_opti
+}
+
+LocalVector<String> EditorBuildProfile::get_build_option_classes(BuildOption p_build_option) {
+	ERR_FAIL_INDEX_V(p_build_option, BUILD_OPTION_MAX, {});
```
```suggestion
	ERR_FAIL_INDEX_V(p_build_option, BUILD_OPTION_MAX, LocalVector<String>());
```

YeldhamDev commented:
@Calinou The PR has been updated with the following:
- Improved the extraction of classes from binary scene files. This removes the need for `Skin` to be hardcoded.
- Added `ADD_CLASS_DEPENDENCY()`. This is used inside `_bind_methods()` so classes can mark if they depend on other classes being registered. This fixes the error with `TabBar`.
- Removed the logic to toggle advanced GUI in the build profile, [as stated by Akien](https://github.com/godotengine/godot/pull/103291#issuecomment-2788794427), that was made before this feature existed, and it's kinda redundant now.

@AThousandShips Changes made.

Calinou commented:
Tested again, it works as expected. The Control Gallery demo now generates the profile correctly.

I get a warning when running the demo with a compiled export template that uses the profile:

```
WARNING: Property not found: 'gui/common/default_scroll_deadzone'.
     at: get_setting_with_override (core/config/project_settings.cpp:404)
```

While rebasing on `master`, to account for the conflict generated by https://github.com/godotengine/godot/commit/55dd5d59e1fd98b0a7904a01cf6c73214556310e, I chose to kept current changes only (i.e. what's in `master`). It performs the check slightly differently, since it checks each MovieWriter individually (which makes more sense to me). It seems to work fine.

akien-mga commented:
I tested the PR on the 2D platformer demo: https://github.com/godotengine/godot-demo-projects/tree/master/2d/platformer

The auto-detection works relatively well, but the `Window` class is wrongly disabled, so the one script fails parsing:
```
SCRIPT ERROR: Parse Error: Native class Window used in script doesn't exist or isn't exposed. [tmp_js_export.js:477:18](http://localhost:8060/tmp_js_export.js)
   at: GDScript::reload (res://game.gd:16)
```

Since the root node is of type `Window`, I believe Window should never be disabled?

Surprisingly, with Window disabled, the 2D platformer demo does work fine though. Only the functionality from the Game script (fullscreen and pause toggles) is broken.

akien-mga commented on editor/editor_build_profile.cpp:
```diff
@@ -40,34 +41,58 @@
-	false, // PHYSICS_2D
-	false, // PHYSICS_3D
-	false, // NAVIGATION
+	false, // NAVIGATION 2D
+	false, // NAVIGATION 3D
```
This is meant to be the suffix after `BUILD_OPTION_` so it should use underscores (same for other occurrences).

```suggestion
	false, // NAVIGATION_2D
	false, // NAVIGATION_3D
```

akien-mga commented on editor/editor_build_profile.cpp:
```diff
@@ -40,34 +41,58 @@
+	false, // PHYSICS_GODOT_3D
+	false, // PHYSICS_JOLT
 	true, // TEXT_SERVER_FALLBACK
 	false, // TEXT_SERVER_COMPLEX
```
```suggestion
	false, // TEXT_SERVER_ADVANCED
```

akien-mga commented on editor/editor_build_profile.cpp:
```diff
@@ -79,38 +104,215 @@ const bool EditorBuildProfile::build_option_disabled_by_default[BUILD_OPTION_MAX
+	false, // PHYSICS_JOLT
 	false, // TEXT_SERVER_FALLBACK
-	false, // TEXT_SERVER_COMPLEX
+	false, // BUILD_OPTION_TEXT_SERVER_ADVANCED
```
`BUILD_OPTION_` is implicit in these comments. (Same for other occurrences.)
```suggestion
	false, // TEXT_SERVER_ADVANCED
```

akien-mga replied:
You missed this one.

akien-mga commented on editor/editor_build_profile.cpp:
```diff
@@ -79,38 +104,215 @@ const bool EditorBuildProfile::build_option_disabled_by_default[BUILD_OPTION_MAX
+			"NavigationRegion3D",
+	} },
+	{ BUILD_OPTION_XR, {
+			"XRBodyModifier3D",
+			"XRHandModifier3D",
+			"XRFaceModifier3D",
+			"XRNode3D",
+			"XROrigin3D",
+			"XRInterface",
+			"XRVRS",
+			"XRBodyTracker",
+			"XRControllerTracker",
+			"XRFaceTracker",
+			"XRHandTracker",
+			"XRInterfaceExtension",
+			"XRPose",
+			"XRPositionalTracker",
+			"XRServer",
+			"XRTracker",
```
Should those be ordered alphabetically so it's easier to review?

akien-mga commented on editor/editor_build_profile.cpp:
```diff
@@ -79,38 +104,215 @@ const bool EditorBuildProfile::build_option_disabled_by_default[BUILD_OPTION_MAX
+			"SpringBoneSimulator3D",
+			"VehicleWheel3D",
+	} },
+	{ BUILD_OPTION_TEXT_SERVER_ADVANCED, {
+			"Control",
```
Does this mean Control is removed if you disable TextServerAdvanced? That seems wrong.

akien-mga commented on editor/editor_build_profile.cpp:
```diff
@@ -79,38 +104,215 @@ const bool EditorBuildProfile::build_option_disabled_by_default[BUILD_OPTION_MAX
+	} },
+};
+
+const HashMap<EditorBuildProfile::BuildOption, LocalVector<String>> EditorBuildProfile::build_option_classes = {
+	{ BUILD_OPTION_3D, {
+			"Node3D",
+	} },
+	{ BUILD_OPTION_NAVIGATION_2D, {
+			"NavigationAgent2D",
+			"NavigationLink2D",
+			"NavigationMeshSourceGeometryData2D",
+			"NavigationObstacle2D"
+			"NavigationPolygon",
+			"NavigationRegion2D",
+	} },
+	{ BUILD_OPTION_NAVIGATION_3D, {
+			"NavigationAgent3D",
+			"NavigationLink3D",
+			"NavigationMeshSourceGeometryData3D",
+			"NavigationObstacle3D",
+			"NavigationRegion3D",
+	} },
+	{ BUILD_OPTION_XR, {
+			"XRBodyModifier3D",
+			"XRHandModifier3D",
+			"XRFaceModifier3D",
+			"XRNode3D",
+			"XROrigin3D",
+			"XRInterface",
+			"XRVRS",
+			"XRBodyTracker",
+			"XRControllerTracker",
+			"XRFaceTracker",
+			"XRHandTracker",
+			"XRInterfaceExtension",
```
Why do we need to list all these explicitly? Don't the SCons options amount to just disabling all these nodes?

YeldhamDev replied:
Because we need to know when any of those are being used to decide if certain features should be disabled or not. For example, even if all nodes related to 2D navigation are disabled, the 2D server for navigation is still compiled unless we tell it otherwise.

akien-mga commented on editor/editor_build_profile.cpp:
```diff
@@ -79,38 +104,215 @@ const bool EditorBuildProfile::build_option_disabled_by_default[BUILD_OPTION_MAX
-	BUILD_OPTION_CATEGORY_GENERAL, // OPENGL
-	BUILD_OPTION_CATEGORY_GENERAL, // VULKAN
+	BUILD_OPTION_CATEGORY_GENERAL, // OPENXR
+	BUILD_OPTION_CATEGORY_GENERAL, // WAYLAND
+	BUILD_OPTION_CATEGORY_GENERAL, // X11
```
These could maybe go in a "Platform" category.

YeldhamDev replied:
I don't think two items are enough to warrant a new category.

akien-mga replied:
For now maybe not, but in the future we might want to expose more options.

E.g. for Linux:
```
        BoolVariable("alsa", "Use ALSA", True),
        BoolVariable("pulseaudio", "Use PulseAudio", True),
        BoolVariable("dbus", "Use D-Bus to handle screensaver and portal desktop settings", True),
        BoolVariable("speechd", "Use Speech Dispatcher for Text-to-Speech support", True),
        BoolVariable("fontconfig", "Use fontconfig for system fonts support", True),
        BoolVariable("udev", "Use udev for gamepad connection callbacks", True),
        BoolVariable("x11", "Enable X11 display", True),
        BoolVariable("wayland", "Enable Wayland display", True),
        BoolVariable("libdecor", "Enable libdecor support", True),
        BoolVariable("touch", "Enable touch events", True),
```
Most of those likely don't have a significant impact on binary size though.

So I guess we can keep it as is for now and revisit later if we add more platform-specific options.

akien-mga commented on editor/editor_build_profile.cpp:
```diff
@@ -192,13 +416,24 @@ String EditorBuildProfile::get_build_option_description(BuildOption p_build_opti
+		TTRC("Forward+ renderer for advanced 3D graphics."),
+		TTRC("Mobile renderer for less advanced 3D graphics."),
 		TTRC("Vulkan back-end of RenderingDevice."),
+		TTRC("Direct3D 12 back-end of RenderingDevice."),
```
This is the only file that uses "back-end" with a hyphen. I'd suggest streamlining it to "backend" which seems more common, since we're changing those strings.

akien-mga commented on editor/editor_build_profile.cpp:
```diff
@@ -79,38 +104,215 @@ const bool EditorBuildProfile::build_option_disabled_by_default[BUILD_OPTION_MAX
+	{ BUILD_OPTION_PHYSICS_JOLT, {
+			BUILD_OPTION_PHYSICS_3D,
+	} },
+	{ BUILD_OPTION_DYNAMIC_FONTS, {
+			BUILD_OPTION_TEXT_SERVER_ADVANCED,
+	} },
```
Is this dependency correct? I think you can use TTF/OTF dynamic fonts with TextServerFallback.

CC @bruvzg

Calinou replied:
Yes, TextServerFallback supports all font file formats that TextServerAdvanced does.

bruvzg replied:
The opposite is also true, `TextServerAdvanced` can support RTL/BiDi with bitmap fonts. So it might be something like `BUILD_OPTION_COMPLEX_TEXT_LAYOUTS`, but not dynamic fonts.

akien-mga commented on editor/editor_build_profile.cpp:
```diff
@@ -405,7 +737,11 @@ void EditorBuildProfileManager::_profile_action(int p_action) {
 		case ACTION_DETECT: {
-			confirm_dialog->set_text(TTR("This will scan all files in the current project to detect used classes."));
+			String text = TTR("This will scan all files in the current project to detect used classes.\nNote that the first scan may take a while, specially in larger projects.");
+#ifdef MODULE_MONO_ENABLED
+			text += "\n\n" + TTR("Warning: Class detection for C# scripts is not currently available, and such files will be ignored.");
+#endif // MODULE_MONO_ENABLED
```
FYI @godotengine/dotnet - would be good to implement as a follow-up.

YeldhamDev replied:
Oh believe me, I tried, and it's currently not possible in a way that makes it worth it. You can ask @raulsntos for more details.

akien-mga commented on scene/gui/color_picker.cpp:
```diff
@@ -1993,6 +1993,10 @@ void ColorPicker::_bind_methods() {
 	BIND_THEME_ITEM_EXT(Theme::DATA_TYPE_STYLEBOX, ColorPicker, mode_button_pressed, "tab_selected", "TabContainer");
 	BIND_THEME_ITEM_EXT(Theme::DATA_TYPE_STYLEBOX, ColorPicker, mode_button_hover, "tab_selected", "TabContainer");
+
+	ADD_CLASS_DEPENDENCY("LineEdit");
+	ADD_CLASS_DEPENDENCY("MenuButton");
+	ADD_CLASS_DEPENDENCY("PopupMenu");
```
How do you determine which classes should be added as dependencies?

I see `color_picker.cpp` has an internal LineEdit but there's also e.g. SpinBox, HSlider, etc.

Is this a requirement only when signals are used?

YeldhamDev replied:
> Is this a requirement only when signals are used?

Seems so, as connecting signals require the class to be registered.

akien-mga replied:
FYI @KoBeWi. I wonder if we could find a more elegant way to do this so we don't need to hardcode such list only for internal classes that have signal connections.

It's fine for now for this PR though.

akien-mga reviewed (commented):
Did a code review pass, looks pretty good overall.

Let's finalize this and get it merged :D

akien-mga commented on servers/register_server_types.cpp:
```diff
@@ -364,12 +363,14 @@ void unregister_server_types() {
 
 	ServersDebugger::deinitialize();
 	memdelete(shader_types);
+
 	if (GD_IS_CLASS_ENABLED(MovieWriterMJPEG)) {
 		memdelete(writer_mjpeg);
 	}
 	if (GD_IS_CLASS_ENABLED(MovieWriterPNGWAV)) {
 		memdelete(writer_pngwav);
 	}
+
```
There's only whitespace changes in this file in the end so this can be reverted.

That should also prevent conflicts with #106013 (which could use an approval btw :wink: ).

akien-mga reviewed (approved):
Looks great to me!
The configuration system isn't perfect yet but this is much better than what was before, and we can build upon it.

I have a couple minor outstanding comments but this should be good to merge soon.

YeldhamDev commented on scene/resources/resource_format_text.cpp:
```diff
@@ -1223,15 +1223,10 @@ Error ResourceLoaderText::get_classes_used(HashSet<StringName> *r_classes) {
-			error_text = "Missing 'type' in external resource tag";
-			_printerr();
-			return error;
+		if (next_tag.fields.has("type")) {
+			r_classes->insert(next_tag.fields["type"]);
```
Yes, instanced scenes.

Repiteo commented on core/io/resource_importer.cpp:
```diff
@@ -456,6 +457,18 @@ void ResourceFormatImporter::get_dependencies(const String &p_path, List<String>
+	List<Ref<ResourceImporter>> valid_importers;
+	get_importers_for_file(p_path, &valid_importers);
+	for (Ref<ResourceImporter> importer : valid_importers) {
+		return importer->get_build_dependencies(p_path, r_dependencies);
```
MSVC gives a warning for unreachable code here, but the fact this is returning at all seems like a relic of a previous iteration. I assume this is instead meant to be:
```suggestion
		importer->get_build_dependencies(p_path, r_dependencies);
```
